go-controller/pkg/ovn/loadbalancer.go

```diff
@@ -190,16 +194,30 @@ func (ovn *Controller) generateACLNameForOVNCommand(lb string, sourceIP string,
 }
 
 func (ovn *Controller) createLoadBalancerRejectACL(lb, sourceIP string, sourcePort int32, proto kapi.Protocol) (string, error) {
+	applyToPortGroup := false
 	ovn.serviceLBLock.Lock()
 	defer ovn.serviceLBLock.Unlock()
 	switches, err := ovn.getLogicalSwitchesForLoadBalancer(lb)
 	if err != nil {
 		return "", fmt.Errorf("error finding logical switch that contains load balancer %s: %v", lb, err)
 	}
 
-	if len(switches) == 0 {
-		klog.V(5).Infof("Ignoring creating reject ACL for load balancer %s. It has no logical switches", lb)
-		return "", nil
+	if len(switches) > 0 {
+		applyToPortGroup = true
+	} else {
+		klog.V(5).Infof("Ignoring creating reject ACL for port group with load balancer %s. It has no "+
+			"logical switches", lb)
+	}
+
+	// check if the load balancer is on a GR, if so we need to get the external switches
+	gwRouterExtSwitch, err := ovn.getGRLogicalSwitchForLoadBalancer(lb)
+	if err != nil {
+		return "", fmt.Errorf("unable to query logical switches for GR with load balancer: %s, error: %v", lb, err)
+	}
+
+	if len(switches) == 0 && gwRouterExtSwitch == "" {
+		return "", fmt.Errorf("load balancer %s does not apply to any switches in the cluster. Will not create "+
+			"Reject ACL", lb)
```

Comment:
Can this actually happen? It looks like this error will propagate up and result in the service being half-added.

Reply:
It's extremely unlikely, but it's not bad to add some code to catch this case. Let's break down what happens for each possible createLoadBalancerRejectACL call:
1. endpoints deletion: if all the endpoints are gone for a service, try to create a reject ACL. If no switches, create no ACL. Load balancer endpoints are still cleared. The result will just be no reject happens when a client tries to access the LB. But that LB is not on any network entity so it is a moot point. Behavior here is OK.
2. service creation: if this errors because the load balancer is not on a switch, then yeah we will end up returning out of the service add. So the other potential svc Ports (if using a different protocol) may not get their reject ACL created. This could be a pitfall if somehow one protocol load balancer was on a switch but another wasn't. If you want I can modify the service add code to handle this case. Let me know what you think.